### ISSUE TITLE
Validation loss keeps increasing

### ISSUE THREAD

**Issue by pranavchat14** [gauravchak/two_tower_models#3]
# Increasing Test Loss Despite Training Convergence

## Observation
Training loss decreases while test loss increases over 5 epochs, with declining retrieval metrics:

```
Epoch [1/5] - Train Loss: 5.1010
Epoch [1/5] - Test Metrics:
loss: 4.8458
hit_rate@10: 0.0650
ndcg@10: 0.0254
mrr: 0.0135
precision@10: 0.0065
recall@10: 0.0650


Epoch [5/5] - Train Loss: 0.8231
Epoch [5/5] - Test Metrics:
loss: 6.2107
hit_rate@10: 0.0450
ndcg@10: 0.0166
mrr: 0.0084
precision@10: 0.0045
recall@10: 0.0450
```

## Configuration
```python
model = TwoTowerBaseRetrieval(
    num_items=num_items_to_return,
    user_id_hash_size=2048,
    user_id_embedding_dim=128,
    user_features_size=43,
    item_id_hash_size=100000,
    item_id_embedding_dim=128,
    item_features_size=388,
    user_value_weights=[1.0],
    mips_module=mips_module,
)
```

## Dataset
Using dataset with real user interactions

## Environment
- Device: CPU
- Evaluation metrics: NDCG@10, MRR, Hit Rate@10, Precision@10, Recall@10
- PyTorch Version: 2.5.1
- Full evaluation code and metrics attached below

<details>
<summary>Full Evaluation Results and Code</summary>

### Complete Metrics Progression
```
Epoch [1/5] - Train Loss: 5.1010
Epoch [1/5] - Test Metrics:
loss: 4.8458
hit_rate@10: 0.0650
ndcg@10: 0.0254
mrr: 0.0135
precision@10: 0.0065
recall@10: 0.0650



Epoch [2/5] - Train Loss: 2.5945
Epoch [2/5] - Test Metrics:
loss: 5.1947
hit_rate@10: 0.0700
ndcg@10: 0.0291
mrr: 0.0173
precision@10: 0.0070
recall@10: 0.0700



Epoch [3/5] - Train Loss: 1.6833
Epoch [3/5] - Test Metrics:
loss: 5.6855
hit_rate@10: 0.0450
ndcg@10: 0.0220
mrr: 0.0151
precision@10: 0.0045
recall@10: 0.0450



Epoch [4/5] - Train Loss: 1.2580
Epoch [4/5] - Test Metrics:
loss: 6.1182
hit_rate@10: 0.0500
ndcg@10: 0.0201
mrr: 0.0113
precision@10: 0.0050
recall@10: 0.0500



Epoch [5/5] - Train Loss: 0.8231
Epoch [5/5] - Test Metrics:
loss: 6.2107
hit_rate@10: 0.0450
ndcg@10: 0.0166
mrr: 0.0084
precision@10: 0.0045
recall@10: 0.0450
```

### Evaluation Implementation
```python
import numpy as np
import torch

def compute_dcg(relevances, k=None):
    """Compute DCG (Discounted Cumulative Gain)."""
    if k is not None:
        relevances = relevances[:k]
    return np.sum(relevances / np.log2(np.arange(2, len(relevances) + 2)))

def compute_ndcg(relevances, k=None):
    """Compute NDCG (Normalized Discounted Cumulative Gain)."""
    dcg = compute_dcg(relevances, k)
    ideal_relevances = np.sort(relevances)[::-1]
    idcg = compute_dcg(ideal_relevances, k)
    return dcg / idcg if idcg > 0 else 0.0

def compute_mrr(relevances):
    """Compute MRR (Mean Reciprocal Rank)."""
    ranks = np.where(relevances == 1)[0]
    if len(ranks) > 0:
        return 1.0 / (ranks[0] + 1)
    return 0.0

def evaluate(model, dataloader, device, k=10):
    """
    Evaluate model performance on the validation/test set.
    Returns dict with metrics: NDCG@k, MRR, Hit Rate@k, etc.
    """
    model.eval()
    total_loss = 0.0
    total_hits = 0
    total_samples = 0
    
    ndcg_scores = []
    mrr_scores = []
    precision_scores = []
    recall_scores = []
    
    with torch.no_grad():
        for batch in dataloader:
            (
                user_ids,
                user_features, 
                user_history,
                item_ids,
                item_features,
                positions,
                labels,
            ) = batch

            # Move to device
            user_ids = user_ids.to(device)
            user_features = user_features.to(device) 
            user_history = user_history.to(device)
            item_ids = item_ids.to(device)
            item_features = item_features.to(device)
            positions = positions.to(device)
            labels = labels.to(device)

            # Get model predictions
            batch_loss = model.train_forward(
                user_ids,
                user_features,
                user_history,
                item_ids, 
                item_features,
                positions,
                labels,
            )

            # Get recommendations
            recommendations = model(
                user_ids,
                user_features,
                user_history
            )

            # Calculate metrics for each user
            for rec, label in zip(recommendations, labels):
                rec = rec.cpu().numpy()
                label = label.cpu().item()
                
                # Create binary relevance vector
                relevances = np.zeros_like(rec)
                relevances[rec == label] = 1
                
                # Calculate metrics
                ndcg = compute_ndcg(relevances, k)
                mrr = compute_mrr(relevances)
                precision = np.mean(relevances[:k])
                
                # Handle division by zero for recall
                total_relevances = np.sum(relevances)
                recall = np.sum(relevances[:k]) / total_relevances if total_relevances > 0 else 0.0
                
                ndcg_scores.append(ndcg)
                mrr_scores.append(mrr)
                precision_scores.append(precision)
                recall_scores.append(recall)
                
                if label in rec[:k]:
                    total_hits += 1
                    
            total_samples += len(labels)
            total_loss += batch_loss.item()

    # Compute final metrics
    metrics = {
        'loss': total_loss / len(dataloader),
        f'hit_rate@{k}': total_hits / total_samples,
        f'ndcg@{k}': np.mean(ndcg_scores),
        'mrr': np.mean(mrr_scores),
        f'precision@{k}': np.mean(precision_scores),
        f'recall@{k}': np.mean(recall_scores)
    }
    
    return metrics
```
</details>

## Labels
- bug
- performance
